### PR TITLE
BLD: Allow numpy 2.x

### PR DESCRIPTION
**Pull request by pllim**
### Description

This pull request is an attempt to fix devdeps that now fails to build Cython because numpy-dev no longer allows building against its 1.x ABI. I think this means we have to *build* astropy with numpy-dev to test against numpy-dev.

Example log with devdeps failure: https://github.com/astropy/astropy/actions/runs/6926199400/job/18837993098

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by pllim** [comment]
@astrofrog , since you fixed a similar problem for astropy dev nightly wheel, do you know how to properly fix this?

**Comment by astrofrog** [comment]
Will look into it shortly!

**Comment by pllim** [comment]
Should we use "no build isolation" for devdeps too? If so, how to do it in a way OpenAstronomy workflow would let us here?

File: astropy/table/tests/test_column.py
Comment (pllim) [inline astropy/table/tests/test_column.py:1]: I have no idea what this change means but this is what I had to do to make it green with numpy 2.0.dev

File: astropy/table/tests/test_column.py
Comment (mhvk) [inline astropy/table/tests/test_column.py:384]: Since `UFuncNoLoopError` is a subclass of `TypeError`, only the match needs to be changed, i.e.,
```
with pytest.raises(TypeError, match="string operation on non-string array" if NUMPY_LT_2_0 else "ufunc 'str_len' did not contain a loop"):
    ...
```

File: astropy/table/tests/test_column.py
Comment (mhvk) [inline astropy/table/tests/test_column.py:368]: To me, this looks like an improvement that the string length of a masked element can no longer influence the result, even if I'm not sure why it has happened! Pinging @taldcroft just in case.

**Review by mhvk — COMMENTED** [review]
The code changes look good to me, modulo the small simplification that can be made.

File: astropy/table/tests/test_column.py
Comment (neutrinoceros) [reply astropy/table/tests/test_column.py:368]: Admittedly Big Endian architecture are pretty rare in the wild, but I'm marginally worried about the endianness being hard-coded here.
```suggestion
            expected_c2_dtype = "U2"
```

Comment (taldcroft) [reply astropy/table/tests/test_column.py:368]: Hmm, I have the opposite view. `MaskedColumn` is built on the premise of being able to unmask values with no data loss and the astropy `insert` method is supposed to respect that. I.e. it shouldn't matter if the inserted data are masked or not, you should later be able to unmask and get the original back.

One guess is that there is a change in `np.char.str_len()` where it now only counts unmasked elements?
https://github.com/astropy/astropy/blob/fc804b88c43afcbd42986d6fda4679a7b6d3a578/astropy/table/column.py#L149

Comment (taldcroft) [reply astropy/table/tests/test_column.py:368]: Yup.
Numpy < 2.0:
```
In [2]: c = np.ma.MaskedArray(["ccc", "dd"], mask=[True, False])

In [3]: np.char.str_len(c).max()
Out[3]: 3

In [4]: np.char.str_len(c)
Out[4]: array([3, 2])
```
Numpy 2.0dev
```
In [2]: c = np.ma.MaskedArray(["ccc", "dd"], mask=[True, False])

In [3]: np.char.str_len(c)
Out[3]: 
masked_array(data=[--, 2],
             mask=[ True, False],
       fill_value=999999)

In [4]: np.char.str_len(c).max()
Out[4]: np.int64(2)
```
I agree that `np.char.str_len()` is doing the right thing in 2.0dev, so we should unmask the values if necessary. I'll have a look at this.

Comment (taldcroft) [reply astropy/table/tests/test_column.py:368]: I think this should work for the referenced line above. In my testing this should unmask the data and be pretty safe to do no matter what `values` is and it doesn't make a copy even for a masked array.
```
        # Starting with numpy 2.0, np.char.str_len() respects the mask for masked data.
        # For astropy we ensure that masked values are preserved so we need to 
        # unmask `values` prior to counting string lengths.
        values = np.asarray(values)

        # Find the length of the longest string in the new values.
        values_str_len = np.char.str_len(values).max()
```

Comment (pllim) [reply astropy/table/tests/test_column.py:368]: @taldcroft , which file is this? I am taking the rest of the week off, so feel free to push to this branch, or I'll revisit next week. Thanks!

Comment (mhvk) [reply astropy/table/tests/test_column.py:368]: @taldcroft - looking at this, I think one solution would be to take the length from the dtype if `values` has one, but probably best to just stick with your solution here, to make the changes minimal. I'll push some changes.

Separately, `np.insert(array, ...)` also converts integer, etc., to string, so we're not quite consistent with that. But to be honest I think it is better to be a bit more strict.

File: astropy/table/tests/test_column.py
Comment (mhvk) [reply astropy/table/tests/test_column.py:384]: I've done this as part of what I'll push in a sec; squashing my changes with your third commit.

**Review by mhvk — APPROVED** [review]
Really nice to work again with numpy 2.0!

**Review by taldcroft — APPROVED** [review]
Looks good, thanks all!

**Comment by pllim** [comment]
@saimn , is this okay to get in between RC2 and release? Or you want to wait?

**Comment by pllim** [comment]
Thanks for cleaning up my PR, y'all!

**Comment by saimn** [comment]
I think for now we should keep the `numpy>=1.25,<2` pin in build requirements, for the v6.0.0 release. Otherwise building it may break once numpy 2.0 is out.

**Comment by saimn** [comment]
Actually for the release we probably also want to pin numpy<2 in runtime requirements.

**Comment by pllim** [comment]
Feel free to re-milestone then. I'll let you merge when you think it is ready. Thanks!

**Comment by pllim** [comment]
@astrofrog is confused about this and I think he is investigating, so probably has to wait for his approval?

@saimn and @astrofrog need to make a final decision whether this goes in v6.0.1 or v6.1.

**Comment by pllim** [comment]
I rebased. Let's see if the devdeps still passes.

**Comment by pllim** [comment]
Of course the CI isn't triggering... GitHub status is yellow right now.

**Comment by pllim** [comment]
devdeps still green. 🤷

**Comment by mhvk** [comment]
Hmm, weird indeed, this branch works fine on my Debian laptop.

**Comment by pllim** [comment]
Hmm how is `py311-test-devdeps -vv -r` different from what we're running here besides the Python version? I really don't understand what tox is doing here. Does this mean my patch isn't the correct fix?

**Comment by mhvk** [comment]
Given the last e-mail on astropy-dev, it is perhaps surprising that things work for me (with tox 4.11.3), so let me copy the output here:
```
py311-test-devdeps: 174047 W commands[0] /home/mhvk/data/mhvk/packages/astropy/.tmp/py311-test-devdeps> pip freeze [tox/tox_env/api.py:427]
astropy @ file:///home/mhvk/data/mhvk/packages/astropy/.tox/.tmp/package/1/astropy-6.1.dev126%2Bg1e1f39039f.tar.gz#sha256=bb6c9354b27a28c867a23202dbbf21458ff79b34ac9faa44bcd3b706abea7df4
astropy-iers-data==0.2023.11.27.0.30.38
attrs==23.1.0
coverage==7.3.2
execnet==2.0.2
hypothesis==6.91.0
iniconfig==2.0.0
numpy==2.0.0.dev0
packaging==23.2
pluggy==1.3.0
pyerfa==2.0.1.2.dev9+g0d5ddca
pytest==7.4.3
pytest-arraydiff==0.6.1
pytest-astropy==0.11.0
pytest-astropy-header==0.2.2
pytest-cov==4.1.0
pytest-doctestplus==1.0.0
pytest-filter-subpackage==0.1.2
pytest-mock==3.12.0
pytest-remotedata==0.4.1
pytest-xdist==3.5.0
PyYAML==6.0.1
SciPy==1.12.0.dev0
sortedcontainers==2.4.0
py311-test-devdeps: 174500 I exit 0 (0.45 seconds) /home/mhvk/data/mhvk/packages/astropy/.tmp/py311-test-devdeps> pip freeze pid=543390 [tox/execute/api.py:279]
py311-test-devdeps: 174500 W commands[1] /home/mhvk/data/mhvk/packages/astropy/.tmp/py311-test-devdeps> pytest --pyargs astropy /home/mhvk/data/mhvk/packages/astropy/docs [tox/tox_env/api.py:427]
=============================================================== test session starts ===============================================================
platform linux -- Python 3.11.6, pytest-7.4.3, pluggy-1.3.0
cachedir: .tox/py311-test-devdeps/.pytest_cache

Running tests with Astropy version 6.1.dev126+g1e1f39039f.
Running tests in astropy docs.

Date: 2023-11-28T08:16:26

Platform: Linux-6.5.0-4-amd64-x86_64-with-glibc2.37

Executable: /home/mhvk/data/mhvk/packages/astropy/.tox/py311-test-devdeps/bin/python

Full Python Version: 
3.11.6 (main, Oct  8 2023, 05:06:43) [GCC 13.2.0]

encodings: sys: utf-8, locale: UTF-8, filesystem: utf-8
byteorder: little
float info: dig: 15, mant_dig: 15

Package versions: 
Numpy: 2.0.0.dev0+git20231124.94bc564
Scipy: 1.12.0.dev0+2025.f016aa9
Matplotlib: not available
h5py: not available
Pandas: not available
PyERFA: 2.0.1.2.dev9+g0d5ddca
Cython: not available
Scikit-image: not available
pyarrow: not available
asdf-astropy: not available

Using Astropy options: remote_data: none.

CI: undefined
ARCH_ON_CI: undefined
IS_CRON: undefined

rootdir: /home/mhvk/data/mhvk/packages/astropy
configfile: pyproject.toml
plugins: astropy-header-0.2.2, arraydiff-0.6.1, filter-subpackage-0.1.2, astropy-0.11.0, remotedata-0.4.1, xdist-3.5.0, cov-4.1.0, hypothesis-6.91.0, mock-3.12.0, doctestplus-1.0.0
collecting ...                                                                                                                             ^
```

**Review by astrofrog — APPROVED** [review]
Ok I finally understand why this works:

```
py312-test-devdeps: remove tox env folder /home/tom/Code/astropy/.tox/py312-test-devdeps
.pkg: remove tox env folder /home/tom/Code/astropy/.tox/.pkg
py312-test-devdeps: install_deps> python -I -m pip install -v --pre 'numpy>=0.0.dev0' 'pyerfa>=0.0.dev0' 'scipy>=0.0.dev0'
.pkg: install_requires> python -I -m pip install 'cython<3.1.0,>=3.0.0' extension-helpers 'numpy>=1.25' setuptools 'setuptools_scm>=6.2'
```
this above is a red herring, even though it installs numpy 1.26 it's not used to build the package, just I think during the sdist generation phase.

```
.pkg: _optional_hooks> python /home/tom/.local/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: get_requires_for_build_sdist> python /home/tom/.local/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
.pkg: build_sdist> python /home/tom/.local/lib/python3.12/site-packages/pyproject_api/_backend.py True setuptools.build_meta
py312-test-devdeps: install_package_deps> python -I -m pip install -v --pre 'PyYAML>=3.13' 'astropy-iers-data>=0.2023.10.30.0.29.53' 'numpy>=1.22' 'packaging>=19.0' 'pyerfa>=2.0' 'pytest-astropy-header>=0.2.1' 'pytest-astropy>=0.10' 'pytest-doctestplus>=0.12' pytest-xdist 'pytest>=7.0'
py312-test-devdeps: install_package> python -I -m pip install -v --pre --force-reinstall --no-deps /home/tom/Code/astropy/.tox/.tmp/package/4/astropy-6.0.dev1146+g1e1f39039f.d20231128.tar.gz
```

This last line is key, the sdist is installed with pip with the --pre command so I think at that point it gets Numpy dev because ``PIP_EXTRA_INDEX_URL`` is set. So customizing ``install_command`` does seem to be the key here.

**Pull request merged**
